server/client.go

```diff
@@ -3567,19 +3567,35 @@ func (c *client) deliverMsg(prodIsMQTT bool, sub *subscription, acc *Account, su
 			// be a different client that will send the reply.
 			srv.trackGWReply(nil, client.acc, reply, c.pa.reply)
 		}
+		// Capture our num messages which was optimistically updated above.
+		// If this is a service import we will check if we failed to deliver.
+		pnm := sub.nm
 		client.mu.Unlock()
 
+		// For service imports, track if we delivered.
+		didDeliver := true
+
 		// Internal account clients are for service imports and need the '\r\n'.
 		start := time.Now()
 		if client.kind == ACCOUNT {
 			sub.icb(sub, c, acc, string(subject), string(reply), msg)
```

Comment:
Would have the changes been too big to make `msgHandler` (`sub.icb`) return a boolean? I am guessing yes, but just checking.